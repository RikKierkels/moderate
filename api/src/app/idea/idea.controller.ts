import {
  Body,
  Controller,
  Delete,
  Get,
  Param,
  Post,
  Put,
  UseInterceptors
} from '@nestjs/common';
import { IdeaService } from './idea.service';
import { Observable } from 'rxjs';
import {
  ApiNotFoundResponse,
  ApiParam,
  ApiResponse,
  ApiTags
} from '@nestjs/swagger';
import { Auth } from '../shared/decorators/auth.decorator';
import { UserId } from '../shared/decorators/user.decorator';
import { first } from 'rxjs/operators';
import { FindOneParams } from '../shared/find-one-params.model';
import { IsAuthorOfIdeaGuard } from '../shared/guards/is-author-of.guard';
import { IdeaCreateDto } from './models/idea-create.dto';
import { IdeaUpdateDto } from './models/idea-update.dto';
import { IdeaWithMessagesDto } from './models/idea-messages.dto';
import { IdeaDto } from './models/idea.dto';
import { MapResponseInterceptor } from '../shared/interceptors/map-response.interceptor';
import Mapper from '../shared/mappers/entity-mapper';
import { IdeaEntity } from '../database/entities/idea.entity';

@ApiTags('Idea')
@Controller('ideas')
export class IdeaController {
  constructor(private readonly ideaService: IdeaService) {}

  @UseInterceptors(new MapResponseInterceptor(Mapper.mapToIdeaDto))
  @ApiResponse({ type: [IdeaDto] })
  @Get()
  findAll(): Observable<IdeaEntity[]> {
    return this.ideaService.findAll$();
  }

  @UseInterceptors(new MapResponseInterceptor(Mapper.mapToIdeaWithMessagesDto))
  @ApiResponse({ type: IdeaWithMessagesDto })
  @ApiNotFoundResponse()
  @ApiParam({ name: 'id', type: FindOneParams })
  @Get(':id')
  find(@Param() params: FindOneParams): Observable<IdeaEntity> {
    return this.ideaService.findById$(params.id);
  }

  @UseInterceptors(new MapResponseInterceptor(Mapper.mapToIdeaDto))
  @ApiResponse({ type: IdeaDto })
  @ApiNotFoundResponse()
  @Auth()
  @Post()
  create(
    @Body() ideaToCreate: IdeaCreateDto,
    @UserId() userId: string
  ): Observable<IdeaEntity> {
    return this.ideaService.create$(ideaToCreate, userId);
  }

  @UseInterceptors(new MapResponseInterceptor(Mapper.mapToIdeaDto))
  @ApiResponse({ type: IdeaDto })
  @ApiNotFoundResponse()
  @Auth(IsAuthorOfIdeaGuard)
  @Put()
  update(@Body() ideaToUpdate: IdeaUpdateDto): Observable<IdeaEntity> {
    return this.ideaService.update$(ideaToUpdate);
  }

  @ApiNotFoundResponse()
  @Auth(IsAuthorOfIdeaGuard)
  @ApiParam({ name: 'id', type: FindOneParams })
  @Delete(':id')
  delete(@Param() params: FindOneParams): void {
    this.ideaService
      .delete$(params.id)
      .pipe(first())
      .subscribe();
  }
}
