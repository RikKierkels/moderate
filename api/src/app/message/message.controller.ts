import {
  Body,
  Controller,
  Delete,
  Param,
  Post,
  Put,
  UseInterceptors
} from '@nestjs/common';
import {
  ApiNotFoundResponse,
  ApiParam,
  ApiResponse,
  ApiTags
} from '@nestjs/swagger';
import { Auth } from '../shared/decorators/auth.decorator';
import { UserId } from '../shared/decorators/user.decorator';
import { Observable } from 'rxjs';
import { IsAuthorOfMessageGuard } from '../shared/guards/is-author-of.guard';
import { MessageService } from './message.service';
import { FindOneParams } from '../shared/find-one-params.model';
import { MessageDto } from './models/message.dto';
import { MessageCreateDto } from './models/message-create.dto';
import { MessageUpdateDto } from './models/message-update.dto';
import { MessageEntity } from '../database/database-entities';
import { MapResponseInterceptor } from '../shared/interceptors/map-response.interceptor';
import Mapper from '../shared/mappers/entity-mapper';

@ApiTags('Message')
@Controller('messages')
export class MessageController {
  constructor(private readonly messageService: MessageService) {}

  @UseInterceptors(new MapResponseInterceptor(Mapper.mapToMessageDto))
  @ApiResponse({ type: MessageDto })
  @Auth()
  @Post()
  create(
    @Body() messageToCreate: MessageCreateDto,
    @UserId() userId: string
  ): Observable<MessageEntity> {
    return this.messageService.create$(messageToCreate, userId);
  }

  @UseInterceptors(new MapResponseInterceptor(Mapper.mapToMessageDto))
  @ApiResponse({ type: MessageDto })
  @ApiNotFoundResponse()
  @Auth(IsAuthorOfMessageGuard)
  @Put()
  update(@Body() messageToUpdate: MessageUpdateDto): Observable<MessageEntity> {
    return this.messageService.update$(messageToUpdate);
  }

  @ApiNotFoundResponse()
  @Auth(IsAuthorOfMessageGuard)
  @ApiParam({ name: 'id', type: FindOneParams })
  @Delete(':id')
  delete(@Param() params: FindOneParams): void {
    this.messageService.delete(params.id);
  }
}
