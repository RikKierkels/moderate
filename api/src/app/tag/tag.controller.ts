import { Controller, Get, UseInterceptors } from '@nestjs/common';
import { ApiResponse, ApiTags } from '@nestjs/swagger';
import { TagService } from './tag.service';
import { Observable } from 'rxjs';
import { TagDto } from './models/tag.dto';
import Mapper from '../shared/mappers/entity-mapper';
import { MapResponseInterceptor } from '../shared/interceptors/map-response.interceptor';
import { TagEntity } from '../database/entities/tag.entity';

@ApiTags('Tag')
@Controller('tags')
export class TagController {
  constructor(private readonly tagService: TagService) {}

  @UseInterceptors(new MapResponseInterceptor(Mapper.mapToTagDto))
  @ApiResponse({ type: [TagDto] })
  @Get()
  findAll(): Observable<TagEntity[]> {
    return this.tagService.findAll$();
  }
}
