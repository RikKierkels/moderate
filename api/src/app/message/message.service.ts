import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { MessageEntity } from '../database/database-entities';
import { Repository } from 'typeorm';
import { forkJoin, from, Observable } from 'rxjs';
import { IdeaService } from '../idea/idea.service';
import { catchError, map, switchMap } from 'rxjs/operators';
import { UserService } from '../user/user.service';
import { Service } from '../shared/service.interface';
import { MessageCreateDto } from './models/message-create.dto';
import { MessageUpdateDto } from './models/message-update.dto';

@Injectable()
export class MessageService implements Service<MessageEntity> {
  constructor(
    @InjectRepository(MessageEntity)
    private readonly repository: Repository<MessageEntity>,
    private readonly ideaService: IdeaService,
    private readonly userService: UserService
  ) {}

  findById$(id: string): Observable<MessageEntity> {
    return from(
      this.repository.findOneOrFail(id, { where: { isDeleted: false } })
    ).pipe(
      catchError(() => {
        throw new NotFoundException(`Cannot find message with id: ${id}.`);
      })
    );
  }

  create$(
    messageToCreate: MessageCreateDto,
    userId: string
  ): Observable<MessageEntity> {
    return forkJoin([
      this.userService.findOrCreate$(userId),
      this.ideaService.findById$(messageToCreate.ideaId)
    ]).pipe(
      map(([user, idea]) => {
        return this.repository.create({
          ...messageToCreate,
          idea,
          author: user
        });
      }),
      switchMap(messageEntity => this.repository.save(messageEntity))
    );
  }

  update$(messageToUpdate: MessageUpdateDto): Observable<MessageEntity> {
    return from(
      this.repository.update(messageToUpdate.id, { text: messageToUpdate.text })
    ).pipe(switchMap(() => this.repository.findOne(messageToUpdate.id)));
  }

  delete(id: string): void {
    this.repository.update(id, { isDeleted: true });
  }
}
