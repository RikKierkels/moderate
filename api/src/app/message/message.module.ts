import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
import { MessageController } from './message.controller';
import { MessageService } from './message.service';
import { UserModule } from '../user/user.module';
import { IdeaModule } from '../idea/idea.module';
import { MessageEntity } from '../database/entities/message.entity';

@Module({
  imports: [IdeaModule, UserModule, TypeOrmModule.forFeature([MessageEntity])],
  controllers: [MessageController],
  providers: [MessageService]
})
export class MessageModule {}
