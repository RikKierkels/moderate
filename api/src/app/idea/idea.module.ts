import { Module } from '@nestjs/common';
import { IdeaService } from './idea.service';
import { IdeaController } from './idea.controller';
import { TypeOrmModule } from '@nestjs/typeorm';
import { TagModule } from '../tag/tag.module';
import { UserModule } from '../user/user.module';
import { IdeaEntity } from '../database/entities/idea.entity';

@Module({
  imports: [TagModule, UserModule, TypeOrmModule.forFeature([IdeaEntity])],
  controllers: [IdeaController],
  providers: [IdeaService],
  exports: [IdeaService]
})
export class IdeaModule {}
