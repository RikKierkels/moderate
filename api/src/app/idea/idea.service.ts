import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { forkJoin, from, Observable } from 'rxjs';
import { catchError, map, switchMap } from 'rxjs/operators';
import { TagService } from '../tag/tag.service';
import { UserService } from '../user/user.service';
import { Service } from '../shared/service.interface';
import { IdeaCreateDto } from './models/idea-create.dto';
import { IdeaUpdateDto } from './models/idea-update.dto';
import { IdeaEntity } from '../database/entities/idea.entity';

@Injectable()
export class IdeaService implements Service<IdeaEntity> {
  whereIsNotDeleted = { where: { isDeleted: false } };

  constructor(
    @InjectRepository(IdeaEntity)
    private readonly repository: Repository<IdeaEntity>,
    private readonly tagService: TagService,
    private readonly userService: UserService
  ) {}

  findAll$(): Observable<IdeaEntity[]> {
    return from(this.repository.find({ ...this.whereIsNotDeleted }));
  }

  findById$(id: string): Observable<IdeaEntity> {
    return from(
      this.repository.findOneOrFail(id, { ...this.whereIsNotDeleted })
    ).pipe(
      catchError(() => {
        throw new NotFoundException(`Cannot find idea with id: ${id}.`);
      })
    );
  }

  create$(ideaToCreate: IdeaCreateDto, userId: string): Observable<IdeaEntity> {
    return forkJoin([
      this.userService.findOrCreate$(userId),
      this.tagService.findByIds$(ideaToCreate.tags)
    ]).pipe(
      map(([user, tags]) => {
        return this.repository.create({
          ...ideaToCreate,
          tags,
          author: user
        });
      }),
      switchMap(ideaEntity => this.repository.save(ideaEntity))
    );
  }

  update$(ideaToUpdate: IdeaUpdateDto): Observable<IdeaEntity> {
    return this.tagService.findByIds$(ideaToUpdate.tags).pipe(
      map(tags => this.repository.create({ ...ideaToUpdate, tags })),
      switchMap(ideaEntity => this.repository.save(ideaEntity)),
      switchMap(ideaEntity => this.repository.findOne(ideaEntity.id))
    );
  }

  delete$(id: string): Observable<IdeaEntity> {
    return this.findById$(id).pipe(
      map(ideaEntity => {
        const messages = ideaEntity.messages.map(message => ({
          ...message,
          isDeleted: true
        }));
        return { ...ideaEntity, messages, isDeleted: true };
      }),
      switchMap(ideaEntity => this.repository.save(ideaEntity))
    );
  }
}
