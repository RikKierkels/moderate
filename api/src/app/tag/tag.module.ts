import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
import { TagController } from './tag.controller';
import { TagService } from './tag.service';
import { TagEntity } from '../database/entities/tag.entity';

@Module({
  imports: [TypeOrmModule.forFeature([TagEntity])],
  controllers: [TagController],
  providers: [TagService],
  exports: [TagService]
})
export class TagModule {}
