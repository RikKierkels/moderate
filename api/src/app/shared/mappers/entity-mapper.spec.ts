import { makeMessage } from '../test-helpers/make-entities.test-utils';
import EntityMapper from './entity-mapper';
import { UserEntity } from '../../database/entities/user.entity';
import { IdeaEntity } from '../../database/entities/idea.entity';
import { TagEntity } from '../../database/entities/tag.entity';
import { MessageEntity } from '../../database/entities/message.entity';

const tagEntity: TagEntity = {
  id: '46518359-049f-49aa-8b4b-e41218674c1e',
  name: 'Angular',
  color: '#000'
};

const userEntity: UserEntity = {
  id: '46518359-049f-49aa-8b4b-e41218674c1e',
  username: 'RikKierkels',
  picture: 'https://www.cool-face.com/123.jpg',
  messages: [],
  ideas: []
};

const messageEntity: MessageEntity = {
  id: '46518359-049f-49aa-8b4b-e41218674c1e',
  text: 'I should map to a DTO',
  createdAt: new Date(2020, 1, 20),
  updatedAt: new Date(2020, 1, 20),
  isDeleted: false,
  author: userEntity,
  idea: null
};

const idea: IdeaEntity = {
  id: '46518359-049f-49aa-8b4b-e41218674c1e',
  title: 'Snapshot Test',
  description: 'I should map to a DTO',
  difficulty: 1,
  createdAt: new Date(2020, 1, 20),
  updatedAt: new Date(2020, 1, 20),
  isDeleted: false,
  author: userEntity,
  messages: [messageEntity, messageEntity],
  tags: [tagEntity]
};

describe('Entity Mapper', () => {
  it('should map a tag entity to a tag DTO', () => {
    const tagDto = EntityMapper.mapToTagDto(tagEntity);

    expect(tagDto).toMatchSnapshot();
  });

  it('should map a user entity to a user DTO', () => {
    const userDto = EntityMapper.mapToUserDto(userEntity);

    expect(userDto).toMatchSnapshot();
  });

  describe('While mapping a message entity to a DTO', () => {
    it('should map the message', () => {
      const messageDto = EntityMapper.mapToMessageDto(messageEntity);

      expect(messageDto).toMatchSnapshot();
    });

    it("should change the text if it's deleted", () => {
      let message = makeMessage();
      message = { ...message, isDeleted: true };

      const messageDto = EntityMapper.mapToMessageDto(message);

      expect(messageDto.text).toBe('This message was deleted by the user.');
    });
  });

  it('should map an idea entity to an idea DTO', () => {
    const ideaDto = EntityMapper.mapToIdeaDto(idea);

    expect(ideaDto).toMatchSnapshot();
  });

  it('should map an idea entity to an idea with messages DTO', () => {
    const ideaDto = EntityMapper.mapToIdeaWithMessagesDto(idea);

    expect(ideaDto).toMatchSnapshot();
  });
});
