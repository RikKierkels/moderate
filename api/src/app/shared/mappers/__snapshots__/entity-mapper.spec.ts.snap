// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entity Mapper While mapping a message entity to a DTO should map the message 1`] = `
Object {
  "author": Object {
    "id": "46518359-049f-49aa-8b4b-e41218674c1e",
    "picture": "https://www.cool-face.com/123.jpg",
    "username": "RikKierkels",
  },
  "createdAt": 2020-02-19T23:00:00.000Z,
  "id": "46518359-049f-49aa-8b4b-e41218674c1e",
  "text": "I should map to a DTO",
}
`;

exports[`Entity Mapper should map a tag entity to a tag DTO 1`] = `
Object {
  "color": "#000",
  "id": "46518359-049f-49aa-8b4b-e41218674c1e",
  "name": "Angular",
}
`;

exports[`Entity Mapper should map a user entity to a user DTO 1`] = `
Object {
  "id": "46518359-049f-49aa-8b4b-e41218674c1e",
  "picture": "https://www.cool-face.com/123.jpg",
  "username": "RikKierkels",
}
`;

exports[`Entity Mapper should map an idea entity to an idea DTO 1`] = `
Object {
  "author": Object {
    "id": "46518359-049f-49aa-8b4b-e41218674c1e",
    "picture": "https://www.cool-face.com/123.jpg",
    "username": "RikKierkels",
  },
  "createdAt": 2020-02-19T23:00:00.000Z,
  "description": "I should map to a DTO",
  "difficulty": 1,
  "id": "46518359-049f-49aa-8b4b-e41218674c1e",
  "messageCount": 2,
  "tags": Array [
    Object {
      "color": "#000",
      "id": "46518359-049f-49aa-8b4b-e41218674c1e",
      "name": "Angular",
    },
  ],
  "title": "Snapshot Test",
}
`;

exports[`Entity Mapper should map an idea entity to an idea with messages DTO 1`] = `
Object {
  "author": Object {
    "id": "46518359-049f-49aa-8b4b-e41218674c1e",
    "picture": "https://www.cool-face.com/123.jpg",
    "username": "RikKierkels",
  },
  "createdAt": 2020-02-19T23:00:00.000Z,
  "description": "I should map to a DTO",
  "difficulty": 1,
  "id": "46518359-049f-49aa-8b4b-e41218674c1e",
  "messages": Array [
    Object {
      "author": Object {
        "id": "46518359-049f-49aa-8b4b-e41218674c1e",
        "picture": "https://www.cool-face.com/123.jpg",
        "username": "RikKierkels",
      },
      "createdAt": 2020-02-19T23:00:00.000Z,
      "id": "46518359-049f-49aa-8b4b-e41218674c1e",
      "text": "I should map to a DTO",
    },
    Object {
      "author": Object {
        "id": "46518359-049f-49aa-8b4b-e41218674c1e",
        "picture": "https://www.cool-face.com/123.jpg",
        "username": "RikKierkels",
      },
      "createdAt": 2020-02-19T23:00:00.000Z,
      "id": "46518359-049f-49aa-8b4b-e41218674c1e",
      "text": "I should map to a DTO",
    },
  ],
  "tags": Array [
    Object {
      "color": "#000",
      "id": "46518359-049f-49aa-8b4b-e41218674c1e",
      "name": "Angular",
    },
  ],
  "title": "Snapshot Test",
}
`;
